src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,487 @@
+        // The Administrator Commissioning cluster is always on the root endpoint.
+        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::CurrentMode::Id);
+    }
+    return Protocols::InteractionModel::Status::Success;
+}
+
+Status Instance::UpdateStartUpMode(DataModel::Nullable<uint8_t> aNewStartUpMode)
+{
+    if (!aNewStartUpMode.IsNull())
+    {
+        if (!IsSupportedMode(aNewStartUpMode.Value()))
+        {
+            return Protocols::InteractionModel::Status::InvalidValue;
+        }
+    }
+    DataModel::Nullable<uint8_t> oldStartUpMode = mStartUpMode;
+    mStartUpMode                                = aNewStartUpMode;
+    if (mStartUpMode != oldStartUpMode)
+    {
+        // Write new value to persistent storage.
+        GetAttributePersistenceProvider()->WriteScalarValue(
+            ConcreteAttributePath(mEndpointId, mClusterId, Attributes::StartUpMode::Id), mStartUpMode);
+        // The Administrator Commissioning cluster is always on the root endpoint.
+        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::StartUpMode::Id);
+    }
+    return Protocols::InteractionModel::Status::Success;
+}
+
+Status Instance::UpdateOnMode(DataModel::Nullable<uint8_t> aNewOnMode)
+{
+    if (!aNewOnMode.IsNull())
+    {
+        if (!IsSupportedMode(aNewOnMode.Value()))
+        {
+            return Protocols::InteractionModel::Status::InvalidValue;
+        }
+    }
+    DataModel::Nullable<uint8_t> oldOnMode = mOnMode;
+    mOnMode                                = aNewOnMode;
+    if (mOnMode != oldOnMode)
+    {
+        // Write new value to persistent storage.
+        GetAttributePersistenceProvider()->WriteScalarValue(ConcreteAttributePath(mEndpointId, mClusterId, Attributes::OnMode::Id),
+                                                            mOnMode);
+        // The Administrator Commissioning cluster is always on the root endpoint.
+        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::OnMode::Id);
+    }
+    return Protocols::InteractionModel::Status::Success;
+}
+
+DataModel::Nullable<uint8_t> Instance::GetStartUpMode() const
+{
+    return mStartUpMode;
+}
+
+DataModel::Nullable<uint8_t> Instance::GetOnMode() const
+{
+    return mOnMode;
+}
+
+uint8_t Instance::GetCurrentMode() const
+{
+    return mCurrentMode;
+}
+
+bool Instance::IsSupportedMode(uint8_t modeValue)
+{
+    for (uint8_t i = 0; i < NumberOfModes(); i++)
+    {
+        uint8_t value;
+        auto err = GetModeValueByIndex(i, value);
+        if (err == CHIP_NO_ERROR && value == modeValue)
+        {
+                return true;
+        }
+    }
+    ChipLogDetail(Zcl, "Cannot find a mode with value %u", modeValue);
+    return false;
+}
+
+Instance::~Instance()
+{
+    ModeBaseAliasesInstanceMap.erase(mClusterId);
+}
+
+std::map<uint32_t, Instance *> GetModeBaseInstances()
```

Comment:
This is returning a copy of the map.  That seems pretty expensive....

Reply:
Fixed [here](https://github.com/project-chip/connectedhomeip/pull/27504/commits/f9f5a66fc570efe192d3f5645f2293c7c4fd26c1).